### ISSUE TITLE
Rock age / type / minerals

### ISSUE THREAD

**Issue by Estevo-Aleixo** [davidson16807/tectonics.js#8]
How would I determine what rock held mineral wise? Could this be added to tectonics.js or is it better to do as an additional layer over top? What research is out there?

**Comment by davidson16807** [comment]
You are correct. This can be added and in fact it was one of the long term directions I'd planned to take. 

I'm not that aware of the research but I can make some guesses. Sedimentary rock associates with regions where erosion has occurred. Igneous associates with subduction zones and continental crust. Same goes for metamorphic, but metamorphic rock could also occur in regions of continental collision. 

You can make further refinements by distinguishing rocks by their constituents. Mafic rock occurs primarily in the ocean and in the lower layer of continental crust. Felsic rock occurs primarily in the top continental crust. Mafic rock could be kicked up onto continental crust by volcanic arcs. This would distinguish things like basalt from 

Harder to model but potentially more promising would be to use the [Goldschmidt classification](https://en.wikipedia.org/wiki/Goldschmidt_classification). I'm not as sure how to go about modeling this. Siderophile elements would probably associate with ultramafic rock, but I'm not sure. Knowing where ultramafic rock exists might then tell you where the valuable mineral deposits are - things like gold, silver, and iron.

**Comment by davidson16807** [comment]
Roadmap:

 * internally track age of crust
 * visualize age in custom shader
 * internally track sediment as it is created during erosion
 * use sediment density to derive soil color
 * internally track sima/sial fraction as it is created during island volcanism
 * use sediment density to derive bedrock color
 * internally track past npp as it relates to fossil fuel reserves 
 * visualize fossil fuel distributions in custom shader
 * igneous/metamorphic rock 
 * lithophile/chalcophile elements

Bonus:
 * use crust age to determine density and foundering of oceanic crust

**Comment by davidson16807** [comment]
2a06a3 adds some means to view the age of crust. This is not yet viewable on the website.

There's a few things that pop out at me from implementing this:

 * The age of oceanic crust is all wrong. Oceanic crust should never get much older than 250 my, as we see [on earth](https://upload.wikimedia.org/wikipedia/commons/e/e7/2008_age_of_oceans_plates.jpg). Any portions of crust that are close to this age should either be nearing a subduction zone or bordering the coasts of an "inner" ocean. This is a long standing issue that has to do with how virtually all non-academic tectonic simulations operate. 

 * Passing things to the gpu is a major bottleneck. The simulation will see significant slow down as a result of this enhancement. This is a simple operation that does not afford much optimization. If worse comes to worse, I may have to reduce model resolution. The ultimate solution may be to switch to a model that is entirely based in the gpu, but this is would virtually rewrite the entire application, and in any case it comes with many, many challenges. 

 * Tracking stuff like age blows up save file sizes. Expect an additional 5KB for each parameter added. This still makes for perfectly reasonable file sizes, but it gets us further and further away from our hope to store model state in a url's querystring.

**Comment by davidson16807** [comment]
OK, I'm still alive. I've been working off and on for over a year to implement this feature and fix the oceanic age problem. 

ab32f6 on the tests/field branch is overhauling the engine to work with raster-based scalar/vector fields. This made it much easier to solve the oceanic age problem and implement rock types. When I say it's an overhaul, I mean it's like a totally different model. I can only think to call it version 2.0. I'm going to write a blog post about it, but there's so much redesign put into this I don't think I'll ever be able to properly document it all. 

079bd1 on the sima-sial branch is building off the tests/field branch to implement basic rock types. The first distinction I need to make is between the felsic/sial/continental rock type and the mafic/sima/oceanic rock type. The "continental" rock type is a conserved quantity. Ideally, it should never be created or destroyed. This means I need to implement these rock types first before I can comfortably reimplement erosion and accretion, otherwise I'd just have to reimplement them again. 

4dc07d on the erosion branch is building off sima-sial. I'm pretty satisfied with how the model is looking. The continental rock type is slowly accumulating for reasons I don't understand, but it occurs slowly enough that it's not noticeable. You would have to simulate longer than 5 billion years to start to notice it. In any case, there's a lot of debate in research whether it's accumulating or staying constant. 

The only reason I'm posting now is because I feel confident in releasing this soon. This whole redesign started out as a test branch where I was just experimenting on things, so I didn't feel the need to talk about it. There were a lot of road blocks along the way, so even when it did start looking promising I'd hit a road block and wouldn't be sure if I'd ever release it.

**Comment by davidson16807** [comment]
All this has yet to be released. The only change I need to make before releasing it is to update the save file parser so it works with sima/sial.

**Comment by davidson16807** [comment]
I've fixed the save file reader/writer. Pushing to prod. Brace for imact...

**Comment by nrkn** [comment]
Exciting stuff!

**Comment by astrographer** [comment]
I'll try to post a review ASAP, but right now I'm having a perplexing problem. The web app works fine generally, but when I download the files and run tectonics.js I have problems. I did set up of apache on my own, which wasn't fun, but not your problem.

The thing is, if I set the display to alt or npp, the screen goes black and the time stops incrementing. As soon as I go back to a satellite view, it starts again from where it stopped and the display is fine. This doesn't happen with the actual web app, but only on a local server.

**Comment by davidson16807** [comment]
Oh dear, that alt/npp thing is a bug. There's a shader error if you pop open developer tools in the browser. Probably something to do with the cleanup I've been doing since the release. I'll look into it. 

What's apache for? Shouldn't need a local server to run it. Just open index.html and it should run.

**Comment by astrographer** [comment]
Yeah. But when I started having the alt-npp issue, I thought the problem might've been in how it looks up files on the filesystem vs how Apache looks things up.

I'm assuming, then, that the version online isn't updated to the latest version on git… For curiosity, what is the last release version?

**Comment by davidson16807** [comment]
Last release is tagged 2.0, 323b648ab087aee54a6454eead79fb2d6b382b86 on the master branch. I've been doing some refactoring and performance optimization since then.

**Comment by davidson16807** [comment]
3030a112b0b0ff616225b1428b7d8110783effd7 fixes the alt/npp issue

**Comment by astrographer** [comment]
Finally got it up, and it’s looking good!

**Comment by davidson16807** [comment]
I'm switching focus back to this issue now that #17 is taken care of and I'm in the headspace to take care of this. Here's what's slated for the next push:

 * represent sial/sima in the model using mass, not thickness

 * represent sediment within the model

I have a branch out that already implements these features. I tried applying fixes for #17 there, but in doing so the model produces unrealistically high elevations. I think I'm going to reimplement these changes on the branch for issue #17.

**Comment by davidson16807** [comment]
I've reimplemented sediment on the sediment2 branch. Looks stable. sediment quantity increases monotonically throughout the simulation, so sometimes sediment basins can reach depths >5km on a well aged planet. That's ridiculous for sediment, but completely normal for sedimentary rock. 

The problem is there is no process that converts sediment back to rock other than the subduction of tectonic plates. So it seems like any model that realistically represents sediment must necessarily represent its conversion to sedimentary and metamorphic rock. 

I see hardly any literature that models the rate of lithification or metamorphosis. I'd love to be corrected, here. Until I can find something, I might use a simple half life model for lithification, and another model that metamorphizes any sedimentary rock that exceeds a certain depth.

**Comment by davidson16807** [comment]
The full rock cycle has been implemented as of dbe766. 

Metamorphosis and lithification are triggered whenever overburden pressure exceeds a threshold. 

This is a very basic implementation and there's a lot missing: 

 * only considers overburden from the top crust, so rock isn't metamorphosed when its subducted. This is the worst oversight and I expect to work on this next.
 * doesn't consider rate of conversion, it is assumed metamorphosis/lithification occurs instantaneously when enough pressure is applied
 * doesn't consider metamorphosis/lithification as a function of temperature
 * doesn't consider the pressure of overhead water, but I think this effect is negligible
 * doesn't track the depth or mixture of rock types. Rock types are always assumed to be at a certain depth, e.g. "sial" always occurs above "sima" and below "sediment". This means you can't represent something like Sedona Arizona, where basaltic lava overlays a layer of sedimentary rock. It's also not possible to represent stratigraphy, like alternating bands of limestone and sandstone. Fixing this would require an extensive change to how crusts are represented.
 * doesn't distinguish sediments by particle size, e.g. there's a sedimentary rock type but there's no way to distinguish "sand" from "clay". As a result of this, there's also no way to distinguish "sandstone" from "shale", or "quartzite" from "schist"
 * doesn't distinguish rocks by composition, e.g. there's a "sedimentary" rock type but there's no way to distinguish "sandstone" from "coal"; there's a "metamorphic" rock type but no way to distinguish "marble" from "slate"
 * doesn't distinguish low grade metamorphism from high grade metamorphism, so no way to distinguish "slate" from "gneiss". High grade metamorphism might contain precious gemstones like garnet
 * "sima" and "sial" are functionally equivalent to "mafic igneous" and "felsic igneous", but there is no way to distinguish intrusive rocks from extrusive rocks, so we can't say whether "felsic igneous" means "granite" or "rhyolite", and we can't say whether "mafic igneous" means "gabbro" or "basalt"
 * doesn't consider rock type when calculating weathering rate - all rock types weather at the same rate. This means no mesas, devil's towers, etc.

There's also been extensive architectural change to support adding arbitrary rock types to crust. All functionality for rock types should be encapsulated in two files: 
 * "Crust.js" stores basic information concerning the rock type, like how to copy rasters or whether they count as sial/sima
 * "TectonicsModeling.js" stores complex functionality, like how the rock transforms when subjected to certain conditions

As part of these architectural changes, crust stores all its rasters in a single ArrayBuffer. This array buffer is converted to base64 in save files. In other words, there's no need to modify JsonSerializer.js as you modify rock types. It also means save files will be broken whenever a new rock type is added, but for now I think it's well worth it. 

The model right now is very stable. Sial-type crust is conserved across at least several billion years, and no quantity ever reaches into the negatives. Getting to this point required changes to the way we integrate deltas onto plates.

**Comment by NJSwetye** [comment]
Wow, this has come such a long way. Is there a way to query a grid from UI?

**Comment by davidson16807** [comment]
@NJSwetye There's no way to query from the user interface. However, you can issue commands from the developer console. Try something like:

```
view.setScalarView(new ScalarWorldView( 
        new HeatmapRasterView( { scaling: true }),
        world => world.lithosphere.total_crust.age
    ))
```
